source/extensions/clusters/dynamic_forward_proxy/cluster.cc

```diff
@@ -49,6 +49,13 @@ void Cluster::onDnsHostAddOrUpdate(
   // We should never get a host with no address from the cache.
   ASSERT(host_info->address() != nullptr);
 
+  // TODO(mattklein123): Right now we allow a DNS cache to be shared between multiple clusters.
+  // Though we have connection/request circuit breakers on the cluster, we don't have any way to
+  // control the maximum hosts on a cluster. It's possible we want to move to a model where the
+  // DNS cache is shared globally, but there is a per-cluster cache wrapper that can use the
+  // global DNS cache, but can have its own individual host limits with TTL to avoid excessive
+  // per-cluster memory use in a multi-tenant environment.
```

Comment:
Thanks for adding this.  The only modification I might suggest, is that I think the max-hosts isn't purely a multi-tenant thing.  It's non-obvious that if you turn off the overall response timeout you're not creating a potential connection stall but instead opening an OOM vector.  I think we could just take out the last sentence and I'd be good :-)